drivers/i2c/i2c_esp32.c

```diff
@@ -276,32 +273,14 @@ static inline void i2c_esp32_reset_fifo(const struct i2c_esp32_config *config)
 	esp32_clear_mask32(I2C_TX_FIFO_RST | I2C_RX_FIFO_RST, reg);
 }
 
-static int i2c_esp32_spin_yield(int *counter)
+static int i2c_esp32_wait_timeout(uint32_t *counter)
 {
-	*counter = *counter + 1;
-
-	if (*counter > I2C_ESP32_TIMEOUT) {
-		return -ETIMEDOUT;
-	}
-
-	if (*counter > I2C_ESP32_SPIN_THRESHOLD) {
-		k_yield();
-	}
-
-	return 0;
-}
-
-static int i2c_esp32_transmit(const struct device *dev)
-{
-	const struct i2c_esp32_config *config = dev->config;
-	struct i2c_esp32_data *data = dev->data;
-
-	/* Start transmission and wait for the ISR to give the semaphore */
-	sys_set_bit(I2C_CTR_REG(config->index), I2C_TRANS_START_S);
-	if (k_sem_take(&data->fifo_sem, K_MSEC(I2C_ESP32_TIMEOUT_MS)) < 0) {
-		return -ETIMEDOUT;
+	if (*counter == 0) {
+		return 1;
 	}
 
+	(*counter)--;
+	k_busy_wait(1);
```

Comment:
Why is it busy wait instead of k_sleep?

Reply:
Using `k_sleep(K_USEC(1))` increases latency from 10us to 100us between byte read call.

Reply:
So where does 1ms (`#define I2C_ESP32_TIMEOUT_USEC  1000`) timeout come from?

>This timeout is not for the whole transfer, only to wait I2C channel to become idle.

Become idle from which state?

>Using k_sleep(K_USEC(1)) increases latency from 10us ....

10us, not 1us? Won't that 1ms timeout become 10ms then?

Reply:
@mniestroj , this 1ms wait time comes from esp-idf 3.0 I2C implementation, where it checks I2C bus line errors (or busy).  
Regarding the `k_sleep`, it works based on system tick, which would not be small as ~1us (which we do have with nop operations or `k_busy_wait()`).